src/lib/distances.jl

```diff
@@ -47,10 +47,10 @@ end
   end
 
 @adjoint function (::Euclidean)(x::AbstractVector, y::AbstractVector)
-  D = x.-y
+  D = x .- y
   δ = sqrt(sum(abs2, D))
   function euclidean(Δ::Real)
-    x̄ = (Δ / δ) .* D
+    x̄ = (Δ / (δ + eps(typeof(δ)))) .* D
```

Comment:
I'm wondering whether it would be better to apply epsilon only for the zero case.
Something like this is in my mind:
```julia
δ = sqrt(sum(abs2, D))
δ = max(δ, eps(typeof(δ)))
```

Reply:
Hmm yeah, that's also a good point. Will this not make a negligible difference once we're away from `δ == 0` though?

Reply:
It will provide the correct value until `δ >= eps(typeof(δ)))` and yes it has a negligible effect for `δ < eps(typeof(δ)))`. My main concern with the current implementation is that for small but not negligible differences (compared to machine epsilon) `δ + eps(typeof(δ)))` can already lead to a significant deviation:

```julia
x = [1.0]; y = [1.0 + 1.0e-12]
D = x .- y
δ = sqrt(sum(abs2, D))

D ./ (max(δ, eps(typeof(δ))))
1-element Array{Float64,1}:
 -1.0

D ./ (δ + eps(typeof(δ)))
1-element Array{Float64,1}:
 -0.9997780244173141
```

Reply:
I see. Will make the change.

Reply:
@molet I've made the change. One of the tests is still failing and doing something weird. It's the only test currnetly enabled. Would you mind taking a look and seeing if I'm doing something obviously silly?

Reply:
Hm, your change LGTM - let me check the failing test.

Reply:
Ok, this is crazy: I have spent almost 2 hours to find why on earth the test is failing - but I still could not figure out.
My suggestion is to replace the current function with something like below as this one works well:
```julia
@adjoint function pairwise(::Euclidean, X::AbstractMatrix, Y::AbstractMatrix; dims=2)
    D, back = pullback((X, Y) -> sqrt.(max.(pairwise(SqEuclidean(), X, Y; dims=dims), eps(promote_type(eltype(X), eltype(Y)))^2)), X, Y)
    return D, function(Δ)
        return (nothing, back(Δ)...)
    end
end
```

Reply:
Thanks for battling through this -- this looks like a good solution to me.